### PR TITLE
Select queued tasks in stimuli, not transitions

### PR DESCRIPTION
**Pull request by gjoseph92**
Before, we selected new queued tasks as part of transitions: when one task exited `processing`, we recommended another task from `queued->processing`. Now, we select queued tasks in response to stimuli: first we do one cycle of transitions as usual, then check if there are open threads, and only then do a second cycle of transitions moving tasks from `queued->processing`.

By moving the `queued->processing` transition after any possible `queued->released` transition, this ensures we won't accidentally overwrite the `queued->released` recommendation (#7396).

See discussion in https://github.com/dask/distributed/issues/7396#issuecomment-1349175707.

Closes https://github.com/dask/distributed/issues/7396, closes https://github.com/dask/distributed/issues/7401, closes https://github.com/dask/distributed/issues/7398.

cc @fjetter @crusaderky

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/tests/test_scheduler.py [hunk 1]
```diff
@@ -371,6 +372,67 @@ def __del__(self):
...
+
+        await async_wait_for(lambda: len(s.tasks) == len(second_batch), 5)
+
+        # Second batch should move up the queue and start processing
+        assert len(s.queued) == len(second_batch) - s.total_nthreads, list(
+            s.queued.sorted()
+        )
+
+        await event.set()
+        await c2.gather(second_batch)
```
Comment (fjetter) [inline distributed/tests/test_scheduler.py:433]: This test feels a bit heavy weight considering that we have a couple of very simple reproducers, see https://github.com/dask/distributed/issues/7396#issue-1494040437

If you feel strongly about this test, fine, but please add the other two very simple reproducers as well. Regardless of all the intricate timings and queuing, etc. The reproducers there should be true regardless of what internals look like

**Comment by fjetter** [comment]
I'm OK with the approach of using a stimulus

File: distributed/tests/test_scheduler.py (hunk 1 above)
Comment (gjoseph92) [reply distributed/tests/test_scheduler.py:433]: I added the reproducer from https://github.com/dask/distributed/issues/7396#issue-1494040437 as well (with an Event to avoid timing issues): https://github.com/dask/distributed/blob/79525791d3d710533571cd49aff450539c949ffe/distributed/tests/test_scheduler.py#L375-L383

This test is for https://github.com/dask/distributed/issues/7401, which is a different issue (and I think maybe the same as https://github.com/dask/distributed/issues/7398). `Client.close` is the only codepath that can cause `client_releases_keys` to be called with multiple keys, which is the case that breaks.

**Comment by gjoseph92** [comment]
Tests are all flaky:
- https://github.com/dask/distributed/issues/6311
- https://github.com/dask/distributed/issues/5275

File: distributed/tests/test_scheduler.py (hunk 1 above)
Comment (gjoseph92) [reply distributed/tests/test_scheduler.py:444]: Added a simple test with `restart` as well (fails on main, reproducer for https://github.com/dask/distributed/issues/7398): https://github.com/dask/distributed/blob/31975e5b44552688d586c12c694d9a4716378cf1/distributed/tests/test_scheduler.py#L386-L394

File: distributed/scheduler.py [hunk 2]
```diff
@@ -4620,6 +4592,42 @@ def update_graph(
...
+
+        Must be called after `check_idle_saturated`; i.e. `idle_task_count` must be up
+        to date.
+        """
+        if not self.queued:
+            return
+        slots_available = sum(
+            _task_slots_available(ws, self.WORKER_SATURATION)
+            for ws in self.idle_task_count
+        )
```
Comment (gjoseph92) [inline distributed/scheduler.py:4616]: This is hardly worth mentioning, but potential for a negligible scheduler performance change here (which may not actually be possible):

Now, every time a task completes, we run `_task_slots_available` on all idle workers. Before, we only ran it on the one worker that just completed a task. In most of those cases, `len(self.idle_task_count)` should be 1, so no difference.

I mention this because `_task_slots_available` does already show up in py-spy profiles of the scheduler (usually around 0.5-1%), because it's already called frequently. This would maybe allow it to be called even more frequently.

But I don't think it can actually run more than it needs to:
* When there are more threads than root tasks, so many workers are idle, `queued` would be empty, so this wouldn't run.
* When there are more root tasks than threads, and we're queuing, at most one worker should ever be idle: as soon as it becomes idle, it gets another queued task and is no longer idle.

    The one exception is `client.close()` releasing many processing tasks at once, while a different client has tasks on the queue. In that case, there's a lot of rescheduling to do, and we do need to look at all the workers that just became idle, so no unnecessary work there either.

tl;dr I don't see a theoretical way for this to be a problem, but I haven't benchmarked or profiled to be sure.

Comment (fjetter) [reply distributed/scheduler.py:4616]: Thanks for highlighting.

Apart from the theoretical analysis, this is also something we can easily optimize if it becomes a problem, e.g. `saturation * nthreads` is not something we need to compute every time and even the `processing - long running` could be replaced with a counter we inc/dec during transitions.

TLDR Not concerned and "fixing" it right now feels premature

**Review by fjetter — APPROVED** [review]
